### PR TITLE
T21194

### PR DESCRIPTION
**Pull request by joaquimrocha**
https://phabricator.endlessm.com/T21194

### CODE REVIEW

File: src/gs-installed-page.c [hunk 1]
```diff
@@ -546,44 +553,74 @@ gs_installed_page_has_app (GsInstalledPage *self,
...
 
-		/* do not to add pending apps more than once. */
-		if (gs_installed_page_has_app (self, app) == FALSE)
-			gs_installed_page_add_app (self, pending, app);
+		if (gs_app_get_state (app) == AS_APP_STATE_AVAILABLE)
+			gs_app_set_state (app, AS_APP_STATE_QUEUED_FOR_INSTALL);
 
-		/* incremement the label */
-		cnt++;
+		++pending_apps_count;
```
Comment (rshuler) [inline src/gs-installed-page.c]: Again, for consistency, I prefer `pending_apps_count++;`, but I can live with it as is if you feel strongly.

File: src/gs-installed-page.c [hunk 2]
```diff
@@ -546,44 +553,74 @@ gs_installed_page_has_app (GsInstalledPage *self,
...
+		if (!g_error_matches (error, GS_PLUGIN_ERROR, GS_PLUGIN_ERROR_CANCELLED))
+			g_warning ("failed to refine pending apps: %s", error->message);
+		goto out;
+	}
 
-	/* add new apps to the list */
-	pending = gs_plugin_loader_get_pending (plugin_loader);
-	for (i = 0; i < gs_app_list_length (pending); i++) {
-		app = gs_app_list_index (pending, i);
+	for (guint i = 0; i < gs_app_list_length (list); ++i) {
```
Comment (rshuler) [inline src/gs-installed-page.c]: Stylistically (and for consistency with the rest of gnome-software), I prefer `i++` (reserving `++i` for those cases where the order of operation actually matters), but I can live with it as is if you feel strongly (or has already been approved with `++i` upstream).

File: lib/gs-plugin-loader.c [hunk 3]
```diff
@@ -1797,23 +1775,18 @@ static void
...
 	g_autoptr(GString) s = NULL;
 	g_autofree gchar *file = NULL;
 
 	s = g_string_new ("");
-	pending_apps = priv->pending_apps;
 	g_mutex_lock (&priv->pending_apps_mutex);
-	for (i = (gint) pending_apps->len - 1; i >= 0; i--) {
-		GsApp *app;
-		app = g_ptr_array_index (pending_apps, i);
-		if (gs_app_get_state (app) == AS_APP_STATE_QUEUED_FOR_INSTALL) {
```
Comment (rshuler) [inline lib/gs-plugin-loader.c]: Is it correct that this check for AS_APP_STATE_QUEUED_FOR_INSTALL is no longer necessary?

File: lib/gs-plugin-loader.c [hunk 4]
```diff
@@ -1916,16 +1890,10 @@ GsAppList *
 gs_plugin_loader_get_pending (GsPluginLoader *plugin_loader)
 {
 	GsPluginLoaderPrivate *priv = gs_plugin_loader_get_instance_private (plugin_loader);
-	GsAppList *array;
-	guint i;
+	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->pending_apps_mutex);
```
Comment (rshuler) [inline lib/gs-plugin-loader.c]: Again, why the change to g_mutex_locker_new rather than just keep the g_mutex_lock / g_mutex_unlock?  And, I suppose it doesn't really matter, but do you intentionally include the gs_app_list_new() within the mutex where it used to be outside?

File: lib/gs-plugin-loader.c [hunk 5]
```diff
@@ -2978,10 +2946,9 @@ gs_plugin_loader_network_changed_cb (GNetworkMonitor *monitor,
 		g_autoptr(GsAppList) queue = NULL;
 		g_mutex_lock (&priv->pending_apps_mutex);
 		queue = gs_app_list_new ();
-		for (guint i = 0; i < priv->pending_apps->len; i++) {
-			GsApp *app = g_ptr_array_index (priv->pending_apps, i);
-			if (gs_app_get_state (app) == AS_APP_STATE_QUEUED_FOR_INSTALL)
```
Comment (rshuler) [inline lib/gs-plugin-loader.c]: Again, is it safe to skip the check for AS_APP_STATE_QUEUED_FOR_INSTALL?

File: lib/gs-plugin-loader.c [hunk 6]
```diff
@@ -1776,14 +1776,13 @@ save_install_queue (GsPluginLoader *plugin_loader)
...
 	gboolean ret;
-	gint i;
 	g_autoptr(GError) error = NULL;
 	g_autoptr(GString) s = NULL;
 	g_autofree gchar *file = NULL;
 
 	s = g_string_new ("");
 	g_mutex_lock (&priv->pending_apps_mutex);
-	for (i = (gint) gs_app_list_length (priv->pending_apps) - 1; i >= 0; i--) {
+	for (guint i = 0; i < gs_app_list_length (priv->pending_apps); ++i) {
```
Comment (rshuler) [inline lib/gs-plugin-loader.c]: Again, might want `i++` for consistency.

File: lib/gs-plugin-loader.c [hunk 7]
```diff
@@ -1700,18 +1700,56 @@ gs_plugin_loader_pending_apps_add (GsPluginLoader *plugin_loader,
...
 	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->pending_apps_mutex);
+	gboolean save_queue = FALSE;
 
 	g_assert (gs_app_list_length (list) > 0);
 	for (guint i = 0; i < gs_app_list_length (list); i++) {
 		GsApp *app = gs_app_list_index (list, i);
-		gs_app_list_remove (priv->pending_apps, app);
+		if (gs_app_get_state (app) != AS_APP_STATE_QUEUED_FOR_INSTALL) {
+			save_queue = TRUE;
+			gs_app_list_remove (priv->pending_apps, app);
```
Comment (rshuler) [inline lib/gs-plugin-loader.c]: Is it okay that the removal is within the check for AS_APP_STATE_QUEUED_FOR_INSTALL, where previously there was no such check?

File: lib/gs-plugin-loader.c [hunk 8]
```diff
@@ -1727,6 +1765,10 @@ gs_plugin_loader_pending_apps_remove (GsPluginLoader *plugin_loader,
 		}
 
 	}
+	if (save_queue) {
+		g_clear_pointer (&locker, g_mutex_locker_free);
+		save_install_queue (plugin_loader);
+	}
 	g_idle_add (emit_pending_apps_idle, g_object_ref (plugin_loader));
```
Comment (rshuler) [inline lib/gs-plugin-loader.c]: Is it okay that this line may be within or outside the mutex depending on the state of save_queue?

File: lib/gs-plugin-loader.c [hunk 9]
```diff
@@ -1781,6 +1781,15 @@ load_install_queue (GsPluginLoader *plugin_loader, GError **error)
 		if (!gs_plugin_loader_run_refine (helper, list, NULL, error))
 			return FALSE;
 	}
+
+	/* add the pending apps now */
+	locker = g_mutex_locker_new (&priv->pending_apps_mutex);
+	for (i = 0; i < gs_app_list_length (list); ++i) {
+		GsApp *app = gs_app_list_index (list, i);
+		g_ptr_array_add (priv->pending_apps, g_object_ref (app));
+		g_debug ("adding pending app %s", gs_app_get_unique_id (app));
```
[the remark this answers is not in the document]
Comment (joaquimrocha) [reply lib/gs-plugin-loader.c]: Isn't this what's happening?

File: lib/gs-plugin-loader.c [hunk 10]
```diff
@@ -1781,6 +1781,15 @@ load_install_queue (GsPluginLoader *plugin_loader, GError **error)
 		if (!gs_plugin_loader_run_refine (helper, list, NULL, error))
 			return FALSE;
 	}
+
+	/* add the pending apps now */
+	locker = g_mutex_locker_new (&priv->pending_apps_mutex);
```
[the remark this answers is not in the document]
Comment (joaquimrocha) [reply lib/gs-plugin-loader.c]: I wanted to lock the whole array while adding apps. And I used the locker because it's safer like an autoptr.

File: src/gs-installed-page.c (hunk 2 above)
Comment (joaquimrocha) [reply src/gs-installed-page.c]: I got used to the `++i` with the excuse that it's more efficient if the compiler didn't optimize it, so I like to use it. Let's hear from upstream, I have rebased + pushed the PR again upstream today.

Comment (rshuler) [reply src/gs-installed-page.c]: As I understand it, all modern compilers will optimize it.  In any case, whatever upstream likes is fine with me.  I just noticed that the style appeared to post-increment elsewhere in gnome-software, so I thought it would be nice to remain consistent, but in the end I really don't care.

File: lib/gs-plugin-loader.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (rshuler) [reply lib/gs-plugin-loader.c]: Yes, that does seem to be the case here.  I don't see why I commented here (especially on a commit revert).

File: lib/gs-plugin-loader.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (rshuler) [reply lib/gs-plugin-loader.c]: Okay.  Since there is not much going on in this loop, it does make sense to have the mutex around the entire block.

File: lib/gs-plugin-loader.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (rshuler) [reply lib/gs-plugin-loader.c]: What!!!  This is crazy...  This conversation shows up on two commits.  My comment is with regard to https://github.com/endlessm/gnome-software/pull/326/commits/416ad38c61158e3448db8e7adf6f4bb50f1970a3, where the g_debug statement was added after the g_ptr_array_add.  (So, I don't think I had tried to comment on the revert commit https://github.com/endlessm/gnome-software/pull/326/commits/22562db749693570232863a8685e65c116e0ab95 after all.)

[the remark this answers is not in the document]
Comment (rshuler) [reply lib/gs-plugin-loader.c]: My comment also applies to https://github.com/endlessm/gnome-software/pull/326/commits/27a5e769746c7ac2dd641f3b01e55d5708dadcb0.  I don't understand how github is associating the comments with the individual commits.

File: lib/gs-plugin-loader.c (hunk 3 above)
Comment (rshuler) [reply lib/gs-plugin-loader.c]: This question still stands, at least with regard to commit https://github.com/endlessm/gnome-software/pull/326/commits/27a5e769746c7ac2dd641f3b01e55d5708dadcb0

File: lib/gs-plugin-loader.c (hunk 5 above)
Comment (rshuler) [reply lib/gs-plugin-loader.c]: And the question still applies here, too.

File: lib/gs-plugin-loader.c (hunk 7 above)
Comment (rshuler) [reply lib/gs-plugin-loader.c]: And this question still applies, too.

File: lib/gs-plugin-loader.c (hunk 8 above)
Comment (rshuler) [reply lib/gs-plugin-loader.c]: And, just to be sure, can you answer this question, too.

File: lib/gs-plugin-loader.c [hunk 11]
```diff
@@ -1729,9 +1767,54 @@ gs_plugin_loader_pending_apps_remove (GsPluginLoader *plugin_loader,
...
+		const gchar *id = names[i];
+
+		if (strlen (id) == 0)
+			continue;
+		if (!as_utils_unique_id_valid (id))
+			continue;
+
+		app = gs_plugin_loader_app_create (plugin_loader, id);
+		gs_app_list_add (priv->pending_apps, app);
+		g_debug ("adding pending app %s", gs_app_get_unique_id (app));
```
Comment (rshuler) [inline lib/gs-plugin-loader.c]: On the final file comparison, here is the debug statement after the action.

File: lib/gs-plugin-loader.c (hunk 8 above)
Comment (joaquimrocha) [reply lib/gs-plugin-loader.c]: As far as I know, this just adds the callback + args to be called with idle priority, so it should return before the function is called, which would unlock the mutex. There are other places where this is called while locked, but I can look into how glib manages all this if needed (not today though).

File: lib/gs-plugin-loader.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (joaquimrocha) [reply lib/gs-plugin-loader.c]: In https://github.com/endlessm/gnome-software/commit/27a5e769746c7ac2dd641f3b01e55d5708dadcb0 I see only one `g_debug` which didn't touch. GitHub (but also GitLab) also seem a bit messy in the reviews system for me, I always get a bit lost and honestly didn't see the other questions you reminded me I had to answer too...

File: lib/gs-plugin-loader.c (hunk 11 above)
Comment (joaquimrocha) [reply lib/gs-plugin-loader.c]: Ah, I see it now. I can change that.

File: lib/gs-plugin-loader.c (hunk 3 above)
Comment (joaquimrocha) [reply lib/gs-plugin-loader.c]: Honestly, since the commit is months old already, I don't remember exactly. I think there were several problems related to the apps not being in this state while dealing with them, and decoupling the state from the queue means we don't have to rely on the state (one less way to get the typical issues we have with g-s).

I'd have to thoroughly retest all this again to give you a better answer (let me know if I should do it).

File: lib/gs-plugin-loader.c (hunk 4 above)
Comment (joaquimrocha) [reply lib/gs-plugin-loader.c]: It's because it's better to use the RAII, and I modified most of the function so I thought it'd be nice to sneak in this change. I will change the order though as you pointed out.

File: lib/gs-plugin-loader.c (hunk 5 above)
Comment (joaquimrocha) [reply lib/gs-plugin-loader.c]: See the lousy answer to the other question.

File: lib/gs-plugin-loader.c (hunk 3 above)
Comment (rshuler) [reply lib/gs-plugin-loader.c]: Okay, as long as you think it was intentional, I'm fine.
